plugins/GSdx/GSFastList.h

```diff
@@ -0,0 +1,162 @@
+};
+
+template <class T, uint16 INITIAL_SIZE = 4>
+class FastList {
+private:
+	// The last Element<T> of the list has next_index equal to 0
+	// m_buffer[0] is the head of the list, and m_buffer has m_size elements
+	Element<T>* m_buffer;
+	uint16 m_size;
+	uint16 m_free_indexes_stack_top;
+	// m_buffer indexes that are free to be used are stacked here [END - 1]
+	uint16* m_free_indexes_stack;
+
+public:
+	FastList() {
+		m_buffer = nullptr;
+		Clear();
+	}
+
+	~FastList() {
+		_aligned_free(m_buffer);
+	}
+
+	void Clear() {
+		// Initialize END
+		m_size = INITIAL_SIZE;
+		// Initialize arrays
+		_aligned_free(m_buffer);
+		m_buffer = (Element<T>*)_aligned_malloc(m_size * sizeof(Element<T>) + (m_size - 1) * sizeof(uint16), 64);
+		m_free_indexes_stack = (uint16*)&m_buffer[m_size];
+		// Initialize head
+		m_buffer[0].data = nullptr;
+		m_buffer[0].next_index = End();
+		m_buffer[0].prev_index = 0;
+		// Stack top is 0, bottom is END - 2
+		m_free_indexes_stack_top = 0;
+		// Index 0 reserved for head
+		for (uint16 i = 0; i < m_size - 1; i++) {
+			m_free_indexes_stack[i] = i + 1;
+		}
+	}
+
+	__forceinline uint16 First() const {
+		return m_buffer[0].next_index;
+	}
+
+	__forceinline uint16 End() const {
+		return 0;
+	}
+
+	__forceinline uint16 Next(const uint16 index) const {
+		return m_buffer[index].next_index;
+	}
+
+	__forceinline const T Data(const uint16 index) const {
+		return m_buffer[index].data;
+	}
+
+	__forceinline bool Full() const {
+		return m_free_indexes_stack_top == m_size - 1;
+	}
+
+	// Insert the element in front of the list and return it's position in m_buffer
+	__forceinline uint16 InsertFront(const T data) {
+		if (Full()) {
+			Grow();
+		}
+		// Pop a free index from the stack
+		const uint16 free_index = m_free_indexes_stack[m_free_indexes_stack_top++];
+		m_buffer[free_index].data = data;
+		ListInsertFront(free_index);
+		return free_index;
+	}
+
+	__forceinline void Erase(const uint16 index) {
+		ListRemove(index);
+		m_free_indexes_stack[--m_free_indexes_stack_top] = index;
+	}
+
+	__forceinline void MoveFront(const uint16 index) {
+		if (First() != index) {
+			ListRemove(index);
+			ListInsertFront(index);
+		}
+	}
+
+private:
+	__forceinline void ListInsertFront(const uint16 index) {
+		// Update prev / next indexes to add m_buffer[index] to the chain
+		Element<T>& head = m_buffer[0];
+		m_buffer[index].prev_index = 0;
+		m_buffer[index].next_index = head.next_index;
+		if (head.next_index != End()) {
+			m_buffer[head.next_index].prev_index = index;
+		}
+		head.next_index = index;
+	}
+
+	__forceinline void ListRemove(const uint16 index) {
+		// Update prev / next indexes to remove m_buffer[index] from the chain
+		const Element<T>& to_remove = m_buffer[index];
+		m_buffer[to_remove.prev_index].next_index = to_remove.next_index;
+		if (to_remove.next_index != End()) {
+			m_buffer[to_remove.next_index].prev_index = to_remove.prev_index;
+		}
+	}
+
+	void Grow() {
+		if (m_size == 65535) {
```

Comment:
`_UI16_MAX` is non-standard AFAIK. Also http://en.cppreference.com/w/cpp/types/numeric_limits could be used, though if `USHRT_MAX` is already included, might be easier to use that.